### DISCUSSION TITLE
Suspending DOUBLEBUFFERING just for few lines of code...

### DISCUSSION THREAD

**Discussion by exdiogene** [fdivitto/FabGL#233] [Ideas]
Is it possible to disable the Double Buffering just for a moment and enabling it afterward?

I saw that there is a function in the library named "BitmappedDisplayController::setDoubleBuffered(bool value)" and I am not able to use it... 

I also saw that there is a value named "m_doubleBuffered" and I am not able to modify it directly... 

I cannot use different call to DisplayController.setResolution() without and with "DOUBLEBUFFERING" because I loose canvas height the second time, it goes from 320 to 274 lines...

Any solution would be greatly appreciated

**Comment by fdivitto** [comment]
The two screen buffers and the DMA chain are created during initialization, so double buffering cannot be disabled.
What I can do is give it the option to suspend it, but that will require quite a few changes.

I'll make a method like "suspendDoubleBuffering(bool)", just give me a few days ...

**Comment by exdiogene** [comment]
It would only be useful if I do not need to call "canvas.swapBuffers()" after each screen modification. I need direct change to the screen.

Do your items in fabui need to call "canvas.swapBuffers()" each time, when you interact with your own menu design?

As an example, I have a menu on the screen and I select ToolWidth, this item get highlighted, the width selection appear and I increment the width to the value needed. Is it possible without redrawing the whole menu and all other things at each steps?

**Reply by fdivitto** [reply]
> It would only be useful if I do not need to call "canvas.swapBuffers()" after each screen modification. I need direct change to the screen.

swapBuffers() will not have to be called when double buffering is suspended, because in that case just one buffer is active (the visible one). So every primitive is drawn on visible buffer (on blanking period of course).

**Comment by fdivitto** [comment]
Please download and install current repo. I've just added suspendDoubleBuffering() method.
So to disable double buffering just execute:

```DisplayController.suspendDoubleBuffering(true);```
...from here double buffering is disabled, don't call Canvas.swapBuffers() here!!!!
```DisplayController.suspendDoubleBuffering(false);```
...from here double buffering is enabled, you have to call Canvas.swapBuffers() to show drawing buffer.

It should work, I've tested it in DoubleBuffer demo, alternatively disabling and enabling double buffering.

Of course, as you know, when you draw without double buffering flickering can happen, especially because you have to fill the background every "changing item" you draw.

**Comment by fdivitto** [comment]
another note: when you suspend double buffering, a primitives queue is created (not necessary with double buffering). This, for default, requires about 10K. You could reduce this queue size with:

```
BitmappedDisplayController::queueSize := 128; // requires about 1KB, 11 bytes per item
```

**Comment by exdiogene** [comment]
Thank you very much for your great work, it is really appreciated! :)

I will test it next week, because of my boards issues...
